Bitmap extensions

feiyanke opened the pull request:
some useful functions of Bitmap:
``` 
Bitmap.toBytes
Bitmap.clip
Bitmap.skew
Bitmap.rotate
```

feiyanke commented:
I signed it!

hendraanggrian commented:
I'm liking `toBytes`, but it really should be called `toByteArray`. Also, I'm not sure if JPG shoud be default format over PNG.

romainguy commented on src/main/java/androidx/graphics/Bitmap.kt:
```diff
@@ -111,3 +119,49 @@ inline fun createBitmap(
+ * @param quality Hint to the compressor, 0-100. 0 meaning compress for small size, 100 meaning compress for max quality.
+ * @return ByteArray
+ */
+fun Bitmap.toBytes(format:CompressFormat = JPEG, quality: Int=100):ByteArray
```
Compressing a bitmap can be done in many ways and compressing into a byte array is just a highly specific use case. I'd rather remove this method.

romainguy commented on src/main/java/androidx/graphics/Bitmap.kt:
```diff
@@ -111,3 +119,49 @@ inline fun createBitmap(
+ * @param height The height.
+ * @return the clipped bitmap
+ */
+fun Bitmap.clip(x:Int, y:Int, width: Int, height: Int):Bitmap
```
I'm not sure this method is necessary. It doesn't save much over the `Bitmap.createBitmap()` call for an operation that might not be very common. At least it should be `inline`, better documented and the return type doesn't need to be specified.

feiyanke replied:
`clip` is of more clear meaning than `createBitmap` for bitmap operation. I added `inline` for all methods and updated docs.

romainguy commented on src/main/java/androidx/graphics/Bitmap.kt:
```diff
@@ -111,3 +119,49 @@ inline fun createBitmap(
+ * @param py  The y coordinate of the pivot point.
+ * @return the skewed bitmap
+ */
+fun Bitmap.skew(kx:Float, ky:Float, px:Float = 0f, py:Float = 0f):Bitmap
```
This is not a common usage, please remove.

feiyanke replied:
I perfer add this method in order to let operations of bitmap more complete

romainguy commented on src/main/java/androidx/graphics/Bitmap.kt:
```diff
@@ -111,3 +119,49 @@ inline fun createBitmap(
+ * @param py      The y coordinate of the pivot point.
+ * @return the rotated bitmap
+ */
+fun Bitmap.rotate(degrees:Int, px: Float, py: Float):Bitmap
```
I'm not entirely convinced by how common this use case is. (Also same comments that I've mentioned on `clip()` apply here).

feiyanke commented:
@hendraanggrian It more make sense to use `toByteArray`, I have fixed it and added `IntRange` for `quality`. PNG is a common use format, but I perfer JPG as it could make smaller size for most bitmaps.

romainguy commented on api/current.txt:
```diff
@@ -23,6 +23,7 @@ package androidx.content {
 
   public final class ContextKt {
     ctor public ContextKt();
+    method public static java.util.List<java.lang.String> getPermissions(android.content.Context);
```
Why this change?

feiyanke replied:
sorry, it's a mistake, I have already removed it

romainguy commented on src/main/java/androidx/graphics/Bitmap.kt:
```diff
@@ -111,3 +116,56 @@ inline fun createBitmap(
+ * @param py  The y coordinate of the pivot point.
+ * @return the skewed bitmap
+ */
+inline fun Bitmap.skew(kx:Float, ky:Float, px:Float = 0f, py:Float = 0f):Bitmap
```
Remove, this is not useful

romainguy commented on src/main/java/androidx/graphics/Bitmap.kt:
```diff
@@ -111,3 +116,56 @@ inline fun createBitmap(
+ * @param py      The y coordinate of the pivot point.
+ * @return the rotated bitmap
+ */
+inline fun Bitmap.rotate(degrees:Float, px: Float, py: Float)
```
Not sure this is usefuul

romainguy commented on src/main/java/androidx/graphics/Bitmap.kt:
```diff
@@ -111,3 +116,56 @@ inline fun createBitmap(
+ * @param height The height.
+ * @return the clipped bitmap
+ */
+inline fun Bitmap.clip(x:Int, y:Int, width: Int, height: Int)
```
You should run klint, your formatting is inconsistent (`x:Int` vs `with: Int` for instance)
I would rather call this method `crop()`

romainguy commented on src/main/java/androidx/graphics/Bitmap.kt:
```diff
@@ -111,3 +116,56 @@ inline fun createBitmap(
+ * @param quality Hint to the compressor, 0-100. 0 meaning compress for small size, 100 meaning compress for max quality.
+ * @return ByteArray
+ */
+inline fun Bitmap.toByteArray(format:CompressFormat = CompressFormat.JPEG, @IntRange(from = 0, to = 100) quality: Int=100)
```
I still don't find compressing to a byte array useful. Compressing directly into a stream seems the more common case

romainguy commented on src/main/java/androidx/graphics/Bitmap.kt:
```diff
@@ -118,17 +119,18 @@ inline fun createBitmap(
+ * @return ByteArrayInputStream
  */
-inline fun Bitmap.toByteArray(
+inline fun Bitmap.toStream(
```
That is not what I meant. What I meant is that there shouldn't be an extension on `Bitmap` to compress to a byte array or a stream, etc. `Bitmap.compress()` already accepts an `OuputStream` which is generic enough to cover all use cases. It shouldn't be specialized here.

romainguy commented on src/main/java/androidx/graphics/Bitmap.kt:
```diff
@@ -141,23 +143,9 @@ inline fun Bitmap.toByteArray(
  * @return the clipped bitmap
  */
-inline fun Bitmap.clip(x: Int, y: Int, width: Int, height: Int) =
+inline fun Bitmap.clip(x: Int, y: Int, width: Int, height: Int): Bitmap =
```
The return type is not needed. Please rename this method to `crop()`.

romainguy commented on src/main/java/androidx/graphics/Bitmap.kt:
```diff
@@ -111,3 +113,30 @@ inline fun createBitmap(
+
+/**
+ * Creates a new bitmap, rotated from this bitmap by [degrees] - the specified number of degrees,
+ * with a pivot point at ([px], [py]). The pivot point is the coordinate that should remain
```
There's no pivot point in the API

feiyanke replied:
sorry for that, i added it.

romainguy replied:
Thanks for the change. The pivot point should probably be set to width/2.0f and height/2.0f by default.

feiyanke replied:
I added the default pivot point.

romainguy commented:
I am not against those two extensions, they are actually kind of useful, but they don't really leverage any feature of Kotlin, beyond being simple extensions. rotate with a default pivot point would.
